modules/aruco/src/aruco.cpp

```diff
@@ -0,0 +1,623 @@
+
+
+        // look for XY size
+        CV_Assert(board.objPoints.size()>0);
+        float minX, maxX, minY, maxY;
+        minX = maxX = board.objPoints[0][0].x;
+        minY = maxY = board.objPoints[0][0].y;
+
+        for(int i=0; i<board.objPoints.size(); i++) {
+            for(int j=0; j<4; j++) {
+                minX = min(minX, board.objPoints[i][j].x);
+                maxX = max(maxX, board.objPoints[i][j].x);
+                minY = min(minY, board.objPoints[i][j].y);
+                maxY = max(maxY, board.objPoints[i][j].y);
+            }
+        }
+
+        float sizeX, sizeY;
+        sizeX = maxX-minX;
+        sizeY = maxY-minY;
+
+        float xReduction = sizeX/float(outNoMargins.cols);
+        float yReduction = sizeY/float(outNoMargins.rows);
+
+        // determine the zone where the markers are placed
+        cv::Mat markerZone;
+        if(xReduction > yReduction) {
+            int nRows = sizeY/xReduction;
+            int rowsMargins = (outNoMargins.rows-nRows)/2;
+            markerZone = outNoMargins.rowRange(rowsMargins, outNoMargins.rows-rowsMargins);
+        }
+        else {
+            int nCols = sizeX/yReduction;
+            int colsMargins = (outNoMargins.cols-nCols)/2;
+            markerZone = outNoMargins.colRange(colsMargins, outNoMargins.cols-colsMargins);
+        }
+
+        // now paint each marker
+        for(int m=0; m<board.objPoints.size(); m++) {
+
+            // transform corners to markerZone coordinates
+            std::vector<cv::Point2f> outCorners;
+            outCorners.resize(4);
+            for(int j=0; j<4; j++) {
+                cv::Point2f p0, p1, pf;
+                p0 = cv::Point2f(board.objPoints[m][j].x, board.objPoints[m][j].y);
+                // remove negativity
+                p1.x = p0.x - minX;
+                p1.y = p0.y - minY;
+                pf.x = p1.x*float(markerZone.cols-1)/sizeX;
+                pf.y = float(markerZone.rows-1) - p1.y*float(markerZone.rows-1)/sizeY;
+                outCorners[j] = pf;
+            }
+
+            // get tiny marker
+            int tinyMarkerSize = 10*dictData.markerSize+2;
+            cv::Mat tinyMarker;
+            dictData.drawMarker(board.ids[m], tinyMarkerSize, tinyMarker);
+
+            // interpolate tiny marker to marker position in markerZone
+            cv::Mat inCorners(4,1,CV_32FC2);
+            inCorners.ptr<cv::Point2f>(0)[0]= Point2f ( 0,0 );
+            inCorners.ptr<cv::Point2f>(0)[1]= Point2f ( tinyMarker.cols,0 );
+            inCorners.ptr<cv::Point2f>(0)[2]= Point2f ( tinyMarker.cols,tinyMarker.rows );
+            inCorners.ptr<cv::Point2f>(0)[3]= Point2f ( 0,tinyMarker.rows );
+
+            // remove perspective
+            cv::Mat transformation = cv::getPerspectiveTransform(inCorners, outCorners);
+            cv::Mat aux;
+            cv::warpPerspective(tinyMarker, aux, transformation, markerZone.size(), cv::INTER_NEAREST, cv::BORDER_CONSTANT, cv::Scalar::all(127));
```

Comment:
127 is a magic number here. Should be a parameter settable from outside or a named constant.